lib/awful/wibox.lua

```diff
@@ -32,77 +32,157 @@ local awfulwibox = { mt = {} }
 
 --- Array of table with wiboxes inside.
 -- It's an array so it is ordered.
-local wiboxes = {}
+local wiboxes = setmetatable({}, {__mode = "v"})
 
---- Get a wibox position if it has been set, or return top.
--- @param wb The wibox
--- @return The wibox position.
-function awfulwibox.get_position(wb)
-    for _, wprop in ipairs(wiboxes) do
-        if wprop.wibox == wb then
-            return wprop.position
+-- Compute the margin on one side
+local function get_margin(w, position, auto_stop)
+    local h_or_w = (position == "top" or position == "bottom") and "height" or "width"
+    local ret = 0
+
+    for _, v in ipairs(wiboxes) do
+        -- Ignore the wiboxes placed after this one
+        if auto_stop and v == w then break end
+
+        if v.position == position and v.screen == w.screen and v.visible then
+            ret = ret + v[h_or_w]
         end
     end
-    return "top"
+
+    return ret
 end
 
---- Put a wibox on a screen at this position.
--- @param wb The wibox to attach.
--- @param position The position: top, bottom left or right.
--- @param screen If the wibox it not attached to a screen, specified on which
--- screen the position should be set.
-function awfulwibox.set_position(wb, position, screen)
-    local area = get_screen(screen).geometry
-
-    -- The "length" of a wibox is always chosen to be the optimal size
-    -- (non-floating).
-    -- The "width" of a wibox is kept if it exists.
-    if position == "right" then
-        wb.x = area.x + area.width - (wb.width + 2 * wb.border_width)
-    elseif position == "left" then
-        wb.x = area.x
-    elseif position == "bottom" then
-        wb.y = (area.y + area.height) - (wb.height + 2 * wb.border_width)
-    elseif position == "top" then
-        wb.y = area.y
-    end
+-- `honor_workarea` cannot be used as it does modify the workarea itself.
+-- a manual padding has to be generated.
+local function get_margins(w)
+    local position = w.position
+    assert(position)
 
-    for _, wprop in ipairs(wiboxes) do
-        if wprop.wibox == wb then
-            wprop.position = position
-            break
-        end
+    local margins = {left=0, right=0, top=0, bottom=0}
+
+    margins[position] = get_margin(w, position, true)
+
+    -- Avoid overlapping wiboxes
+    if position == "left" or position == "right" then
+        margins.top    = get_margin(w, "top"   )
+        margins.bottom = get_margin(w, "bottom")
     end
+
+    return margins
 end
 
---- Reset all wiboxes positions.
-local function update_all_wiboxes_position()
-    for _, wprop in ipairs(wiboxes) do
-        awfulwibox.set_position(wprop.wibox, wprop.position, wprop.screen)
+-- Create the placement function
+local function gen_placement(position, stretch)
+    local maximize = (position == "right" or position == "left") and
+        "maximize_vertically" or "maximize_horizontally"
+
+    return placement[position] + (stretch and placement[maximize] or nil)
```

Comment:
Let's assume `stretch` is false. This does e.g. `placement.left + nil`. I don't think `nil` in an addition ever works, does it?